### PR TITLE
[test][SageMaker] add multi region capability for PyTorch SM test

### PR DESCRIPTION
**Pull request by tejaschumbalkar**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true` or `neuron_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: dlc_developer_config.toml [hunk 1]
```diff
@@ -13,18 +13,18 @@ benchmark_mode = false
 
 [build]
 # Frameworks for which you want to disable both builds and tests
-skip_frameworks = []
+skip_frameworks = ["tensorflow", "mxnet", "huggingface_tensorflow", "huggingface_pytorch"]
```
Comment (tejaschumbalkar) [inline dlc_developer_config.toml:16]: temp changes

File: test/sagemaker_tests/pytorch/__init__.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    It uploads image to n_virginia region and return image uri
+    """
+    image_repo_uri, image_tag = ecr_image.split(":")
+    _, image_repo_name = image_repo_uri.split("/")
+    target_image_repo_name = f"{image_repo_name}"
+    regional_ecr_image = reupload_image_to_test_ecr(ecr_image, target_image_repo_name, region)
+    return regional_ecr_image
+
+
+def invoke_pytorch_helper_function(pdx_ecr_image, sagemaker_region, helper_function, helper_function_args):
```
Comment (jeet4320) [inline test/sagemaker_tests/pytorch/__init__.py:153]: func description is missing

File: test/test_utils/__init__.py [hunk 3]
```diff
@@ -75,6 +75,7 @@
 
 PUBLIC_DLC_REGISTRY = "763104351884"
 
+SAGEMAKER_EXECUTION_REGION = ["us-west-2", "us-east-1"]
```
Comment (jeet4320) [inline test/test_utils/__init__.py:78]: lets add the other region as well which we discussed

Comment (jeet4320) [reply test/test_utils/__init__.py:78]: nit: should be  `SAGEMAKER_EXECUTION_REGIONS`

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_training_smdebug.py [hunk 4]
```diff
@@ -18,12 +18,12 @@
...
 from ...integration import training_dir, smdebug_mnist_script, DEFAULT_TIMEOUT
 from ...integration.sagemaker.timeout import timeout
-
+from . import invoke_pytorch_estimator
 
 @pytest.mark.integration("smdebug")
 @pytest.mark.model("mnist")
 @pytest.mark.skip_py2_containers
-def test_training_smdebug(sagemaker_session, framework_version, ecr_image, instance_type):
+def test_training_smdebug(framework_version, ecr_image, sagemaker_region, instance_type):
```
Comment (jeet4320) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_training_smdebug.py:26]: nit: should be `sagemaker_regions`

File: test/sagemaker_tests/pytorch/__init__.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    RETRY = 2
```
Comment (jeet4320) [inline test/sagemaker_tests/pytorch/__init__.py:155]: if tests fail, we need to retry anyway. So let keep RETRY=3

File: test/sagemaker_tests/pytorch/__init__.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    DELAY = 300
```
Comment (jeet4320) [inline test/sagemaker_tests/pytorch/__init__.py:156]: I think at least  `DELAY = 600`

File: test/sagemaker_tests/pytorch/__init__.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    for _ in range(RETRY):
+        for region in sagemaker_region:
+            sagemaker_session = get_sagemaker_session(region)
+            ecr_image = get_ecr_image(pdx_ecr_image, region) if region != "us-west-2" else pdx_ecr_image
```
Comment (jeet4320) [inline test/sagemaker_tests/pytorch/__init__.py:160]: can we make more dynamic for any region passed?
Where is pdx_ecr_image defined? i can't find it in PR changes by search

File: test/sagemaker_tests/pytorch/__init__.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    username, password = get_ecr_login_boto3(target_ecr_client, target_account_id, target_region)
+    save_credentials_to_file(ECR_PASSWORD_FILE_PATH, password)
+    subprocess.check_output(f"cat {ECR_PASSWORD_FILE_PATH} | docker login -u {username} --password-stdin https://{target_account_id}.dkr.ecr.{target_region}.amazonaws.com && docker push {target_image_uri}", shell=True, executable="/bin/bash")
+    delete_file(ECR_PASSWORD_FILE_PATH)
+
+    return target_image_uri
+
+def get_ecr_image(ecr_image, region):
+    """
+    It uploads image to n_virginia region and return image uri
```
Comment (jeet4320) [inline test/sagemaker_tests/pytorch/__init__.py:144]: typo: n_virginia

**Review by jeet4320 — COMMENTED** [review]
LG, lets see how CI runs

**Review by jeet4320 — APPROVED** [review]

**Comment by jeet4320** [comment]
I see SM remote tests are failing

**Comment by tejaschumbalkar** [comment]
SM test local run was green. There was a timeout on one of the test, increased it in the latest commit.

File: test/sagemaker_tests/pytorch/__init__.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    if account_id:
+        query["registryId"] = account_id
+    try:
+        ecr_client.describe_repositories(**query)
+    except ecr_client.exceptions.RepositoryNotFoundException as e:
+        return False
+    return True
+
+def delete_file(file_path):
+    subprocess.check_output(f"rm -rf {file_path}", shell=True, executable="/bin/bash")
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/__init__.py:76]: nit: why not use invoke or directly use a python stdlib method?

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py [hunk 10]
```diff
@@ -11,3 +11,62 @@
...
+                        pytorch.disable_profiler = True
+
+                if upload_s3_data_args:
+                    training_input = upload_s3_data(pytorch, **upload_s3_data_args)
+                    inputs = {'training': training_input}
+
+                pytorch.fit(inputs=inputs, job_name=job_name)
+                return pytorch, sagemaker_session
+
+            except Exception as e:
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/__init__.py:67]: nit: why not
```
except sagemaker.exceptions.UnexpectedStatusException as e:
    if "CapacityError" in str(e):
        continue
    else:
        raise
```

File: test/sagemaker_tests/pytorch/__init__.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    """
+    Helper function to reupload an image owned by a another/same account to an ECR repo in this account to given region, so that
+    this account can freely run tests without permission issues.
+
+    :param source_image_uri: str Image URI for image to be tested
+    :param target_image_repo_name: str Target image ECR repo name
+    :param target_region: str Region where test is being run
+    :return: str New image URI for re-uploaded image
+    """
+    ECR_PASSWORD_FILE_PATH = os.path.join("/tmp", f"{get_unique_name_from_tag(source_image_uri)}.txt")
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/__init__.py:107]: nit: this is styled like a global, can prob just be normal lowercase variable

File: test/sagemaker_tests/pytorch/__init__.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        .replace(image_account_id, target_account_id)
+    )
+
+    client = boto3.client('ecr', region_name = image_region)
+    username, password = get_ecr_login_boto3(client, image_account_id, image_region)
+    save_credentials_to_file(ECR_PASSWORD_FILE_PATH, password)
+
+    # using ctx.run throws error on codebuild "OSError: reading from stdin while output is captured".
+    # Also it throws more errors related to awscli if in_stream=False flag is added to ctx.run which needs more deep dive
+    subprocess.check_output(f"cat {ECR_PASSWORD_FILE_PATH} | docker login -u {username} --password-stdin https://{image_account_id}.dkr.ecr.{image_region}.amazonaws.com && docker pull {source_image_uri}", shell=True, executable="/bin/bash")
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/__init__.py:132]: I think invoke is being installed already as test dep, so you can probably use it

File: test/sagemaker_tests/pytorch/__init__.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    RETRY = 2
+    DELAY = 300
+    for _ in range(RETRY):
+        for region in sagemaker_regions:
+            sagemaker_session = get_sagemaker_session(region)
+            ecr_image = get_ecr_image(ecr_image, region) if region != "us-west-2" else ecr_image
+            try:
+                helper_function(ecr_image, sagemaker_session, **helper_function_args)
+                return
+            except Exception as e:
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/__init__.py:176]: why not:

```
except sagemaker.exceptions.UnexpectedStatusException as e:
    if "CapacityError" in str(e):
        continue
    else:
        raise
```

File: test/sagemaker_tests/pytorch/__init__.py (hunk 9 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py:76]: The function already exist [here](https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/utils/__init__.py#L60), moved a folder above. 
I will take this into consideration when i cleanup the [function](https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/utils/__init__.py#L60) in subsequent PR.

File: test/sagemaker_tests/pytorch/__init__.py (hunk 11 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py:107]: Same here. The function already exist [here](https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/utils/__init__.py#L126), moved a folder above. 
I will take this into consideration when i cleanup the [function](https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/utils/__init__.py#L126) in subsequent PR.

File: test/sagemaker_tests/pytorch/__init__.py (hunk 12 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py:132]: Same here, will add the change in subsequent PR

File: test/sagemaker_tests/pytorch/__init__.py (hunk 5 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py:167]: Retrying for 3 times will lead to timeout in the sagemaker test function as well as codebuild job if there are more ICE seen.

File: test/sagemaker_tests/pytorch/__init__.py (hunk 6 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py:168]: Same here.

File: test/sagemaker_tests/pytorch/__init__.py (hunk 13 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/__init__.py]: this looks good as well, adding the change.

**Comment by jeet4320** [comment]
still SM local timeout issue? Is it related to this PR, if not lets merge the PR as SM remote tests passed.

**Review by jeet4320 — APPROVED** [review]

**Review by arjkesh — APPROVED** [review]

**Pull request merged**